File: src/lib/types/tmdb.ts
export type SearchResults<T> = {
  page: number;
  total_pages: number;
  total_results: number;
  results: T[];
};

export type MovieResult = {
  id: number;
  title: string;
  release_date: string | null;
  overview: string | null;
  poster_path: string | null;
  vote_average: number;
};

export type PersonResult = {
  id: number;
  name: string;
  profile_path: string | null;
  known_for_department: string | null;
};

export type CastMember = {
  id: number;
  name: string;
  character: string | null;
  profile_path: string | null;
  order: number;
};

export type MovieDetails = {
  id: number;
  title: string;
  release_date: string | null;
  overview: string | null;
  poster_path: string | null;
  runtime: number | null;
  vote_average: number;
  genres: { id: number; name: string }[];
  credits: { cast: CastMember[] };
};

export type MovieCredit = {
  id: number;
  title: string;
  release_date: string | null;
  character: string | null;
  poster_path: string | null;
  popularity: number | null;
  vote_average: number | null;
};

export type PersonDetails = {
  id: number;
  name: string;
  biography: string | null;
  profile_path: string | null;
  birthday: string | null;
  place_of_birth: string | null;
  movie_credits: { cast: MovieCredit[] };
};

export const TMDB_IMAGE_BASE = 'https://image.tmdb.org/t/p';

export function tmdbImage(path: string | null, size: 'w200' | 'w500' | 'original' = 'w200'): string | null {
  if (!path) return null;
  return `${TMDB_IMAGE_BASE}/${size}${path}`;
}

export function releaseYear(date: string | null | undefined): string {
  if (!date) return '';
  return date.slice(0, 4);
}

File: src-tauri/src/models/tmdb.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResults<T> {
    pub page: u32,
    pub total_pages: u32,
    pub total_results: u32,
    pub results: Vec<T>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MovieResult {
    pub id: u32,
    pub title: String,
    pub release_date: Option<String>,
    pub overview: Option<String>,
    pub poster_path: Option<String>,
    pub vote_average: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersonResult {
    pub id: u32,
    pub name: String,
    pub profile_path: Option<String>,
    pub known_for_department: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MovieDetails {
    pub id: u32,
    pub title: String,
    pub release_date: Option<String>,
    pub overview: Option<String>,
    pub poster_path: Option<String>,
    pub runtime: Option<u32>,
    pub vote_average: f32,
    pub genres: Vec<Genre>,
    pub credits: Credits,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersonDetails {
    pub id: u32,
    pub name: String,
    pub biography: Option<String>,
    pub profile_path: Option<String>,
    pub birthday: Option<String>,
    pub place_of_birth: Option<String>,
    pub movie_credits: MovieCredits,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Credits {
    pub cast: Vec<CastMember>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CastMember {
    pub id: u32,
    pub name: String,
    pub character: Option<String>,
    pub profile_path: Option<String>,
    pub order: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MovieCredits {
    pub cast: Vec<MovieCredit>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MovieCredit {
    pub id: u32,
    pub title: String,
    pub release_date: Option<String>,
    pub character: Option<String>,
    pub poster_path: Option<String>,
    #[serde(default)]
    pub popularity: Option<f32>,
    #[serde(default)]
    pub vote_average: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Genre {
    pub id: u32,
    pub name: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum MediaType {
    Movie,
    Person,
}
